### ISSUE TITLE
Bug in CDLGBR implementation

### ISSUE THREAD

**Issue by ghost** [hercules-390/hyperion#29]
Neale Ferguson seems to have found a bug in the SoftFloat package's "normalizeRoundAndPackFloat32" and "normalizeRoundAndPackFloat64" functions:

Refer to the thread "Bug in CDLGBR implementation?" on the main Hercules list for details.

Below is Neale's patch:

```
--- /root/hercules/spinhawk/softfloat/softfloat.c 2014-07-29 14:38:13.419776940 -0400
+++ softfloat.c 2014-08-03 09:46:17.099658958 -0400
@@ -405,11 +405,14 @@
normalizeRoundAndPackFloat32( flag zSign, int16 zExp, bits32 zSig )
{
int8 shiftCount;

shiftCount = countLeadingZeros32( zSig ) - 1;
- return roundAndPackFloat32( zSign, zExp - shiftCount, zSig<<shiftCount );
+ if (shiftCount >= 0)
+ return roundAndPackFloat32( zSign, zExp - shiftCount, zSig<<shiftCount );
+ else
+ return roundAndPackFloat32( zSign, zExp + 1, zSig>>1 );

}

/*----------------------------------------------------------
| Returns the fraction bits of the double-precision floating-point value `a'.
@@ -566,12 +569,15 @@
static float64
normalizeRoundAndPackFloat64( flag zSign, int16 zExp, bits64 zSig )
{
int8 shiftCount;

- shiftCount = countLeadingZeros64( zSig ) - 1;
- return roundAndPackFloat64( zSign, zExp - shiftCount, zSig<<shiftCount );
+ shiftCount = countLeadingZeros64(zSig) - 1;
+ if (shiftCount >= 0)
+ return roundAndPackFloat64( zSign, zExp - shiftCount, zSig<<shiftCount );
+ else
+ return roundAndPackFloat64( zSign, zExp + 1, zSig>>1 );

}

#ifdef FLOATX80
```

**Comment by srorso** [comment]
Turns out this can be addressed without implementing PFPO.  Code for issue #33 is being tested; the CDLGBR emulation has been successfully validated on hardware.

**Comment by srorso** [comment]
Addressed and validated as corrected by commit 83f60bf.
